File: Algo/Easy/0100_Same_Tree/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// 2021-04-26
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null || q == null) {
            return p == q;
        } else {
            if (p.val == q.val) {
                return (isSameTree(p.left, q.left) && isSameTree(p.right, q.right));
            } else {
                return false;
            }
        }
    }
}

File: Algo/Easy/0100_Same_Tree/Solution.py
from typing import *
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

## 2022-10-16 (stack-based)
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        nodes_to_check = [(p, q)]
        while nodes_to_check:
            node1, node2 = nodes_to_check.pop()
            if node1 and node2:
                if node1.val != node2.val:
                    return False
                
                nodes_to_check.append((node1.left, node2.left))
                nodes_to_check.append((node1.right, node2.right))
            elif not node1 and not node2:
                continue
            else:
                return False
        
        return True
            